Introduction to Kubernetes Operators (#43)
okteto/blog-proposals: introduction-to-kubernetes-operators.md

date: "2021-12-28"
author: "Joydip Kanjilal"
description: "This article provides an introduction to Kubernetes Operators, why they are needed, their components, operator framework and how it works as well as a discussion on custom operators and why they are needed."

---

# **Introduction to Kubernetes Operators**

When we deploy our application on Kubernetes, we use a variety of Kubernetes objects such as role, ingress, service, deployment, config map, etc. As our application becomes more complicated and our requirements non-generic, it becomes challenging to manage it using native Kubernetes objects. You would often have to use manual intervention or some form of automation to compensate.

The good news is that you've Kubernetes Operators to help. It can handle all these tasks for you, making it easier than ever to deploy new applications on your Kubernetes cluster.

This article talks about Kubernetes Operators, why they are useful, how they work and the best practices in using them.

## **Prerequisites**

This article assumes that you're aware of the basics of Kubernetes. A basic knowledge of Kubernetes operators can be helpful but is not desirable.

> ## **What is Kubernetes?**

Reviewer: I think we could skip this section and simply point to the Kubernetes for Developers series we have under the "Prerequisites" sections, WDYT?

Kubernetes Operators make these operations seamless, consistent, repetitive, and scalable by removing complicated and challenging manual activities, minimizing errors, and enabling optimal application performance.

Built on top of the fundamental Kubernetes concepts namely, resource and controller, it incorporates domain or application-specific knowledge to automate the complete life cycle of the application it manages.

Kubernetes Operators can standardize and simplify complex processes making them consistent, repeatable, and scalable, hence eliminating the need of time-consuming manual tasks and errors.

A Kubernetes Operator is capable of:

-   Installing applications with the necessary configurations

-   Scaling up or scaling down applications

-   Performing automatic backups

-   Initiating upgrades

-   Recovering from failures

-   Performing any additional administrative tasks that can be specified using
    code

## **Stateful and Stateless Applications**

Kubernetes is adept at managing and scaling a wide variety of stateless applications such as web applications, APIs and so on without requiring any knowledge on how these applications work.

> Kubernetes doesn't have the additional domain-specific understanding needed for managing and scaling stateful applications such as monitoring systems. Stateful applications are more complex, and they need custom management. Here’s where Kubernetes Operators comes in.

Reviewer: Maybe give some examples here of some stateful applications like you did for stateless applications?

Kubernetes Operators are adept at managing and automating the lifetime of stateful applications. They can help in managing complex configuration details required for a stateful applications.

## **Why does Kubernetes need Operators?**

Kubernetes Operators are required to automate processes that are otherwise accomplished manually by IT operations. They can help ease the process of managing Kubernetes resources by automating lifecycle management and deployment of software assets. Operators are extensions to Kubernetes that leverage custom resources to manage a Kubernetes application and its components.

An operator makes the application easier to manage so that you can focus on using the application without being concerned about managing it. An Operator oversees all that is needed to guarantee that the application runs seamlessly. Kubernetes operators are adept at backing data, recovering from failures, and automatically upgrading the application over time.

## **Kubernetes Operator Components**

> A Kubernetes Operator comprises the following components:

Reviewer: Ummm I'm a little skeptical about this. There have been similar debates in the past on whether an "operator" needs to manage a "cr" or not for it be called an operator: [ref](https://github.com/kubernetes/website/pull/30469/files#r748962017). Could you maybe update this to something like this?

```suggestion
Usually, you'll see a Kubernetes Operator comprising the following components:
```

**A Custom Resource (CR) -** A custom resource (CR), an extension of the Kubernetes API, allows us to declare our own complex types. In other words, a custom resource is a kind of object that extends the Kubernetes API or enables you to integrate your own API into a project or cluster.

A resource refers to an endpoint in Kubernetes that enables you to store any API objects of a particular kind. For example, Kubernetes' built-in pods include a collection of Pod objects.

**A Custom Resource Definition (CRD) -** A Custom Resource Definition (CRD) extends Kubernetes capabilities, enables you to create a custom resource, and makes it declarative using a custom controller. Like other Kubernetes resources, the CRD is defined in YAML.

The Kubernetes API server handles CRDs the same way it treats any other resource. It then reports the configuration content of a CRD to any authorized Kubernetes API consumer.

**A Controller -** A controller is a control loop that keeps an eye on the state of your cluster and makes or requests changes as needed. The controllers build control loops that constantly compare the intended state of a cluster to its actual state.

### **Reconciliation Loops**

> In Kubernetes, a controller is a software loop also known as the “Reconcile loop”. A controller manager manages the Kubernetes cluster by running controllers in a reconciliation loop in the control plane. It executes a control loop that allows each controller to run by executing its Reconcile() function.

Reviewer: ```suggestion
In Kubernetes, a controller is a software loop also known as the “Reconcile loop”. A controller manager manages the Kubernetes cluster by running controllers in a reconciliation loop in the control plane.
```

Since we aren't really showing any code, let's leave this line out to make it less confusing?

Each controller tries to get the present cluster state closer to the state that is desired. If the current state of the cluster does not correspond to the intended state, the controller takes appropriate action to correct the problem.

![Figure 1](introduction-to-kubernetes-operators-images\Figure 1.png)

Figure 1: The Reconciliation Loop

> Kubernetes Operators take advantage of an infinite loop that cycles through three stages, namely, "Observe," "Diff," and "Act".

Reviewer: Since these words are arbitrary maybe use "Compare" instead?
```suggestion
Kubernetes Operators take advantage of an infinite loop that cycles through three stages, namely, "Observe," "Compare," and "Act".
```

The Operator Framework comprises:

-   Operator SDK: This is the core component of the Operator framework and helps launch an Operator project quickly. It allows developers to create operators based on their expertise without having to understand the intricacies of the Kubernetes API. The Operator SDK offers higher level APIs and abstraction thus saving developers time spent on drilling down into Kubernetes APIs. Instead, the developers can focus more on writing operational logic.
    
-   Operator Lifecycle Manager: Administers the installation, upgrades, and management of all operators operating in a Kubernetes cluster. Operator Lifecycle Manager is a tool that allows you to manage, update, and install all operators, as well their dependencies.
    
-   Operator Metering: Allows operators who offer specialized services to report usage. Operator metering captures historical usage data of a cluster and then generates usage statistics that breakdown consumption by pod or namespace over arbitrary periods of time.

## **How Does a Kubernetes Operator Work?**

> In a Kubernetes cluster, an operator comprises a collection of one or more custom resources, as well as a control loop process, all of which are contained inside a pod and runs inside it.

Reviewer: ```suggestion
In a Kubernetes cluster, an operator manages one or more K8s resources, as well as a control loop process, all of which are contained inside a pod and runs inside it.
```

> When a user edits a custom resource, the operator application detects the change and takes appropriate action. These actions can often be calls to the Kubernetes API.

Reviewer: ```suggestion
When a user edits a resource, the operator application detects the change and takes appropriate action. These actions can often be calls to the Kubernetes API.
```

![Figure 2](introduction-to-kubernetes-operators-images\Figure 2.png)

Figure 2: Kubernetes Operator at work!

> Operators use controllers to monitor the behavior and actions of Kubernetes objects. These differ from regular controllers in that they track custom objects, also called custom resource definitions (CRDs).

Reviewer: ```suggestion
Operators use controllers to monitor the behavior and actions of Kubernetes objects. These differ from regular controllers in that they track custom objects, also called custom resources.
```

> The CRD is an extension of the Kubernetes API allowing you to store structured data representing the desired application state. Operators keep track of cluster events that relate to specific types or custom resources. An operator can monitor events that are added, updated, or deleted.

Reviewer: Could you make the definition of CRDs a bit more in line with what's here: https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/#customresourcedefinitions? We should make it clear that CRDs are just ways of defining custom resources.

As soon as the operator observes changes in the environment, the custom controller takes measures to bring the cluster to the desired state. That is known as the "reconciliation loop".

Here's a summary of the workflow:

> 1.  A Custom Resource Definition (CRD) is modified by the user

Reviewer: ```suggestion
1.  A K8s object (could be a custom resource or an existing kubernetes object) is modified by the user
```

2.  The operator monitors the CRD and detects changes

3.  The operator compares the current CRD state to the desired state

4.  Finally, the operator reconciles the cluster state to the desired state

> ## Steps to Build an Operator

Reviewer: I think either we should go in a bit more detial about what the differences are b/w the three or we can just skip this section all together.

You can build an Operator in several ways:

-   Operator SDK - this is a framework that contains helper functions to build operators in Ansible, Go, or HEML.
    
-   Kubebuilder - this represents a framework for building Kubernetes APIs

-   ClientGo - this can connect to the Kubernetes API, but the downside is that it has a steep learning curve.

> ## Deploy an Operator

Reviewer: Same here, else we could just mention this in a line instead of a seperate section.

You can deploy an Operator in the following ways:

-   Using Helm chart

-   Using Yaml as the Kubernetes manifest

> ## **Why Do We Need Custom Operators?**

Reviewer: Could you merge this with "Why does Kubernetes need Operators?" section above?

There are several benefits to implementing your custom Operators:

-   Automate operational tasks that are often handled manually

-   Makes applications more Kubernetes-friendly

-   Assists in testing all scenarios

-   Human mistakes are eliminated

-   Reduces the complexity of application management in hybrid and multi-cloud environments

> ## **Best Practices for Writing Kubernetes Operators**

Reviewer: Could you please elaborate a bit on each of the bullets so the reader knows why we're saying what we're saying?
